mam/v1/merkle.c

```diff
@@ -38,15 +35,26 @@ size_t merkle_depth(size_t const node_count) {
 static size_t merkle_node_index_traverse(size_t const acc, size_t const depth,
                                          size_t const width,
                                          size_t const tree_depth) {
-  if (depth == 0) return acc;
-  if (width < (1 << (depth - 1))) {
-    return merkle_node_index_traverse(acc + 1, depth - 1, width,
-                                      tree_depth - 1);
-  } else {
-    return merkle_node_index_traverse(
-        1 + acc + binary_tree_size(0, tree_depth - 1), depth - 1,
-        width - (1 << (depth - 1)), tree_depth - 1);
+  if (!tree_depth)
+    return 0;
+
+  int depth_cursor = 1;
+  int width_of_leaves_divide_scale = 1;
+  int index = 0;
+  int width_cursor = width;
+  int width_of_leaves = 1 << depth;
+
+  while (depth_cursor <= depth) {
+    if (width_cursor >= (width_of_leaves >> width_of_leaves_divide_scale)) {
```

Comment:
Instead of shifting `1` to the left then to the right, just shift `1` to the left of `depth - depth_cursor`